firewall.tf

```diff
@@ -0,0 +1,66 @@
+resource "google_compute_firewall" "elasticsearch_allow_healthchecks" {
+  name     = "elasticsearch-gcp-health-check"
+  network  = var.network
+  priority = 1000
+
+  source_ranges = [
+    # all the subnets for GCP health checks
+    "35.191.0.0/16", "130.211.0.0/22", "35.191.0.0/16", "209.85.152.0/22", "209.85.204.0/22"
+  ]
+
+  allow {
+    protocol = "tcp"
+  }
+
+  target_tags = ["elasticsearch"]
+  direction   = "INGRESS"
+}
+
+resource "google_compute_firewall" "elasticsearch_allow_ilb_traffic" {
+  name     = "elasticsearch-gce-traffic"
+  network  = var.network
+  priority = 1000
+
+  source_ranges = [
+    "192.168.0.0/16",  # internal load balancer subnet used for traffic towards GCE instances
+  ]
+
+  allow {
+    protocol = "tcp"
+  }
+
+  target_tags = ["elasticsearch"]
+  direction   = "INGRESS"
+}
+
+resource "google_compute_firewall" "elasticsearch_allow_external_subnets" {
+  for_each = length(var.allowed_ipv4_subnets) > 0 ? {allowed_ipv4_subnets: var.allowed_ipv4_subnets} : {}
```

Comment:
If you change `allowed_ipv4_subnets` type to `set`, you can use simple `for_each = var.allowed_ipv4_subnets`. If there is some reason for the `list` type, I think you can use `for_each = toset(var.allowed_ipv4_subnets)`.